### PR TITLE
[WIP] Scale bandSize

### PR DESCRIPTION
**Pull request by felixcodes**
Part of #2. 

Add basic support for scale.clamp, scale.exponent, scale.round. Add scaleType() helper method used for inferring correct scale type from encoding's type. Includes tests for scale properties and scaleType helper.

### CODE REVIEW

**Comment by felixcodes** [comment]
@kanitw Scale bandSize is no longer WIP.

**Comment by kanitw** [comment]
Test is failing.

File: src/config.ts [hunk 1]
```diff
@@ -98,6 +100,7 @@ export const DEFAULT_QUERY_CONFIG: QueryConfig = {
   numberOrdinalProportion: .05,
 
   maxBinsList: [5, 10, 20],
+  scaleBandSizes: [0, 21, 90],
```
Comment (kanitw) [inline src/config.ts]: why 90?

File: src/constraint/spec.ts [hunk 2]
```diff
@@ -76,6 +78,26 @@ export interface SpecConstraint extends AbstractConstraint {
   satisfy: SpecConstraintChecker;
 }
 
+export function getTrueScaleType(encQ: EncodingQuery): ScaleType {
```
Comment (kanitw) [inline src/constraint/spec.ts]: rename to just `scaleType`

Comment (kanitw) [inline src/constraint/spec.ts]: Make sure to add unit test for this.

Comment (kanitw) [inline src/constraint/spec.ts]: This should be in `query.ts` and soon move to `query/encoding.ts`

File: src/constraint/spec.ts [hunk 3]
```diff
@@ -532,9 +554,30 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
     }
   },
   {
+     name: 'scaleBandSizeMustMatchScaleType',
```
Comment (kanitw) [inline src/constraint/spec.ts]: Our last conversation, we mentiont that we only want one `scalePropertiesSupportedByScaleType` rather than having one for each scale property.   Please jot down something we already agree on.  It's not fun to have to keep repeating something for you.   

Plus, why is this a spec constraint?  (It seems like an encoding constraint for me.)

File: src/config.ts (hunk 1 above)
Comment (kanitw) [inline src/config.ts]: 0 is totally unacceptable default value for this `scale.bandSize`. 

Why on the earth would people want a band size to be zero?

File: src/query.ts [hunk 4]
```diff
@@ -345,3 +346,23 @@ export function stringifyEncodingQueryFieldDef(encQ: EncodingQuery): string {
...
+    } else if (type === Type.ORDINAL || type === Type.NOMINAL) {
+      return ScaleType.ORDINAL;
+    } else if (type === Type.TEMPORAL) {
+      if (encQ.timeUnit) {
+        return defaultScaleType(encQ.timeUnit as TimeUnit);
+      } else {
+        return ScaleType.TIME;
+      }
+    }
+  }
```
Comment (kanitw) [inline src/query.ts]: This method does not always return a value.  Does it pass compilation at all?

File: src/property.ts [hunk 5]
```diff
@@ -160,3 +170,110 @@ export function getNestedEncodingPropertyChild(parent: Property) {
 export function isNestedEncodingProperty(prop: Property) {
   return prop in NESTED_ENCODING_INDEX;
 }
+
+const SUPPORTED_SCALE_PROPERTY =
```
Comment (kanitw) [inline src/property.ts]: Why is this added in this PR?  Do you use it at all?

File: src/query.ts [hunk 6]
```diff
@@ -345,3 +346,23 @@ export function stringifyEncodingQueryFieldDef(encQ: EncodingQuery): string {
     params.map((p) => ',' + p.key + '=' + p.value).join('');
   return (fn ? fn + '(' + fieldType + ')' : fieldType);
 }
+
+export function scaleType(encQ: EncodingQuery): ScaleType {
```
Comment (kanitw) [inline src/query.ts]: Don't take the whole `encQ` as input. Take `scaleType`, `timeUnit`, and `type` as inputs so when you call it from constraint, it's clear that you only use only these properties of `encQ`.

File: src/constraint/encoding.ts [hunk 7]
```diff
@@ -221,7 +221,7 @@ export const ENCODING_CONSTRAINTS: EncodingConstraintModel[] = [
     strict: true,
     satisfy: (encQ: EncodingQuery, schema: Schema, opt: QueryConfig) => {
       if (encQ.scale) {
-        const scaleType = (encQ.scale as ScaleQuery).type;
+        const scaleType = actualScaleType(encQ);
```
Comment (kanitw) [inline src/constraint/encoding.ts]: I don't like `actualScaleType` ... maybe name the variable `_scaleType` ?

File: src/constraint/spec.ts [hunk 8]
```diff
@@ -323,7 +324,14 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
     if (mark === Mark.AREA || mark === Mark.BAR) {
       for (let encQ of encodings) {
         if((encQ.channel === Channel.X || encQ.channel === Channel.Y) && encQ.scale) {
-          if (((encQ.scale as ScaleQuery).type as ScaleType) === ScaleType.LOG) {
+
+          const scale: ScaleQuery = encQ.scale as ScaleQuery;
+          let sType = scale.type as ScaleType;
```
Comment (kanitw) [inline src/constraint/spec.ts]: I don't think as need declarations for `sType, timeUnit, type` here as we only use them once.  Just pass them to the method.

File: src/constraint/spec.ts [hunk 9]
```diff
@@ -323,7 +324,14 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
+          let timeUnit = encQ.timeUnit as TimeUnit;
+          let type = encQ.type as Type;
+          let scaleType = _scaleType(sType, timeUnit, type);
```
Comment (kanitw) [inline src/constraint/spec.ts]: Umm, please read my comments carefully.   I mentioned that "use `_scaleType` as local variable and keep method name `scaleType`.

(That said, `sType` maybe better than `_scaleType`.)

File: src/constraint/spec.ts [hunk 10]
```diff
@@ -534,16 +542,22 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
     strict: true,
     satisfy: (specM: SpecQueryModel, schema: Schema, opt: QueryConfig) => {
       const encodings = specM.getEncodings();
 
       for (let encQ of encodings) {
         if (encQ.scale) {
+
           const scale: ScaleQuery = encQ.scale as ScaleQuery;
-          if (contains([ScaleType.LOG, ScaleType.ORDINAL, ScaleType.TIME, ScaleType.UTC], scale.type) &&
+          let sType = scale.type as ScaleType;
```
Comment (kanitw) [inline src/constraint/spec.ts]: unnecessary declaration

File: src/query/encoding.ts [hunk 11]
```diff
@@ -48,3 +48,22 @@ export function isMeasure(encQ: EncodingQuery) {
...
+  if (scaleType !== undefined) {
+    return scaleType;
+  }
+
+  if (type === Type.QUANTITATIVE) {
+    return ScaleType.LINEAR;
+  } else if (type === Type.ORDINAL || type === Type.NOMINAL) {
+    return ScaleType.ORDINAL;
+  } else {
+    // Down here, type is guaranteeed to be Type.TEMPORAL
```
Comment (kanitw) [inline src/query/encoding.ts]: > type is guaranteeed to be Type.TEMPORAL

How do you assume this?
  
This is not true.  Type can be enumSpec in this case.  Not to mention that we might support more type in the future.   

Check for temporal, check for enum spec and throw error for else case.

File: src/query/encoding.ts [hunk 12]
```diff
@@ -48,3 +48,22 @@ export function isMeasure(encQ: EncodingQuery) {
...
+    if (timeUnit !== undefined) {
+      return defaultScaleType(timeUnit);
```
Comment (kanitw) [inline src/query/encoding.ts]: Does this behave correctly with enum spec?

File: src/query/encoding.ts [hunk 13]
```diff
@@ -48,3 +48,22 @@ export function isMeasure(encQ: EncodingQuery) {
   return (encQ.type === Type.QUANTITATIVE && !encQ.bin) ||
       (encQ.type === Type.TEMPORAL && !encQ.timeUnit);
 }
+
+export function scaleType(scaleType: ScaleType, timeUnit: TimeUnit, type: Type) {
```
Comment (kanitw) [inline src/query/encoding.ts]: This method doesn't have unit test yet.

File: src/query/encoding.ts (hunk 12 above)
Comment (kanitw) [inline src/query/encoding.ts]: (If so, make sure to have 100% test coverage for this method.)

File: src/constraint/encoding.ts [hunk 14]
```diff
@@ -221,8 +222,10 @@ export const ENCODING_CONSTRAINTS: EncodingConstraintModel[] = [
     strict: true,
     satisfy: (encQ: EncodingQuery, schema: Schema, opt: QueryConfig) => {
       if (encQ.scale) {
-        const scaleType = (encQ.scale as ScaleQuery).type;
-        const type = encQ.type;
+        const sType = (encQ.scale as ScaleQuery).type as ScaleType;
+        const timeUnit = encQ.timeUnit as TimeUnit;
+        const type = encQ.type as Type;
+        const scaleType = _scaleType(sType, timeUnit, type);
```
Comment (kanitw) [inline src/constraint/encoding.ts]: 1) Unnecessary declarations, why don't you pass these parameters directly to the methods?
2) I commented that use _scaleType as local variable (In hindsight, `sType` might be better) and keep the method called `scaleType`.

File: src/query/encoding.ts [hunk 15]
```diff
@@ -48,3 +48,38 @@ export function isMeasure(encQ: EncodingQuery) {
...
       (encQ.type === Type.TEMPORAL && !encQ.timeUnit);
 }
+
+/**
+ *  Returns the true scale type of an encoding.
+ *  @returns {ScaleType} If the scale type was not specified, it is inferred from the encoding's Type.
+ *  @returns {undefined} If the scale type was not specified and Type (or TimeUnit if applicable) is an EnumSpec, there is no clear scale type
+ */
+
+export function scaleType(scaleType: ScaleType, timeUnit: TimeUnit | EnumSpec<TimeUnit> | ShortEnumSpec, type: Type | EnumSpec<Type> | ShortEnumSpec) {
```
Comment (kanitw) [inline src/query/encoding.ts]: Why the other two parameters can be enum spec but scaleType cannot be enum spec? 

I wonder if you can actually simplify this method to never accept enum spec? 
(Im not sure about that you need to check all reference if you always have them in specific form.)

File: src/constraint/spec.ts [hunk 16]
```diff
@@ -323,7 +324,12 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
     if (mark === Mark.AREA || mark === Mark.BAR) {
       for (let encQ of encodings) {
         if((encQ.channel === Channel.X || encQ.channel === Channel.Y) && encQ.scale) {
-          if (((encQ.scale as ScaleQuery).type as ScaleType) === ScaleType.LOG) {
+
+          let sType = scaleType((encQ.scale as ScaleQuery).type as ScaleType,
+                                    encQ.timeUnit as TimeUnit,
```
Comment (kanitw) [inline src/constraint/spec.ts]: tabbing not aligned.

File: src/query/encoding.ts [hunk 17]
```diff
@@ -48,3 +48,38 @@ export function isMeasure(encQ: EncodingQuery) {
...
+    if (timeUnit !== undefined) {
+      if (isEnumSpec(timeUnit)) {
+        return undefined;
+      }
+      return defaultScaleType(timeUnit as TimeUnit);
+    } else {
+      return ScaleType.TIME;
+    }
+  } else {
+    /* istanbul ignore next */
```
Comment (kanitw) [inline src/query/encoding.ts]: Is this truly the right location for ignore next?  Does test coverage say that this entire else branch is ignored?   (I doubt it should be one line above.)

File: src/constraint/spec.ts [hunk 18]
```diff
@@ -543,8 +549,10 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
       for (let encQ of encodings) {
         if (encQ.scale) {
           const scale: ScaleQuery = encQ.scale as ScaleQuery;
-          if (contains([ScaleType.LOG, ScaleType.ORDINAL, ScaleType.TIME, ScaleType.UTC], scale.type) &&
-             (scale.zero === true)) {
+
+          if (contains([ScaleType.LOG, ScaleType.ORDINAL, ScaleType.TIME, ScaleType.UTC],
+              scaleType(scale.type as ScaleType, encQ.timeUnit as TimeUnit, encQ.type as Type)) &&
```
Comment (kanitw) [inline src/constraint/spec.ts]: assign `const sType = ...`   This line is getting too long.  (But no need for declaration for each parameter.)

File: src/model.ts [hunk 19]
```diff
@@ -54,6 +54,9 @@ export interface EnumSpecIndex {
   /** List of indice tuples of encoding mappings that require scale enumeration. */
   scale?: EnumSpecIndexTuple<boolean>[];
 
+  /** List of indice tuple for encoding mappings that require enumerating scale.scale_bandsize */
```
Comment (kanitw) [inline src/model.ts]: scale.bandSize

File: src/model.ts [hunk 20]
```diff
@@ -54,6 +54,9 @@ export interface EnumSpecIndex {
...
+  scaleBandSize?: EnumSpecIndexTuple<number>[];
+
   /** List of indice tuple for encoding mappings that require enumerating scale.scale_type */
```
Comment (kanitw) [inline src/model.ts]: scale.type

File: src/query/encoding.ts (hunk 15 above)
Comment (kanitw) [inline src/query/encoding.ts]: Thinking more about it, just make it accept enum spec so you can eliminate all lengthy `as` casting in all of usage of this function.

File: src/constraint/encoding.ts [hunk 21]
```diff
@@ -222,24 +221,22 @@ export const ENCODING_CONSTRAINTS: EncodingConstraintModel[] = [
     strict: true,
     satisfy: (encQ: EncodingQuery, schema: Schema, opt: QueryConfig) => {
       if (encQ.scale) {
-        const sType = (encQ.scale as ScaleQuery).type as ScaleType;
-        const timeUnit = encQ.timeUnit as TimeUnit;
         const type = encQ.type as Type;
-        const scaleType = _scaleType(sType, timeUnit, type);
+        const sType = scaleType((encQ.scale as ScaleQuery).type, encQ.timeUnit, encQ.type as Type);
```
Comment (kanitw) [inline src/constraint/encoding.ts]: why still `as Type` here?

File: src/constraint/spec.ts [hunk 22]
```diff
@@ -326,8 +326,7 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
         if((encQ.channel === Channel.X || encQ.channel === Channel.Y) && encQ.scale) {
 
           let sType = scaleType((encQ.scale as ScaleQuery).type as ScaleType,
```
Comment (kanitw) [inline src/constraint/spec.ts]: why still using these as since you already correct `scaleType` 's signature?

File: test/query/encoding.test.ts [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+      const sType = scaleType(undefined, SHORT_ENUM_SPEC, Type.TEMPORAL);
+      assert.equal(sType, undefined);
+    });
+
+    it('should return ScaleType.TIME if type is temporal and scale type and TimeUnit are not specified', () => {
+      const sType = scaleType(undefined, undefined, Type.TEMPORAL);
+      assert.equal(sType, ScaleType.TIME);
+    });
+
+    [TimeUnit.HOURS, TimeUnit.DAY, TimeUnit.MONTH, TimeUnit.QUARTER].forEach((timeU) => {
```
Comment (kanitw) [inline test/query/encoding.test.ts]: why `timeU`  -- we use `timeUnit` everywhere

File: test/query/encoding.test.ts [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+      it('should return ScaleType.ORDINAL if type is temporal and TimeUnit is ' + timeU + ' and scale type is not specified', () => {
+        const sType = scaleType(undefined, timeU, Type.TEMPORAL);
+        assert.equal(sType, ScaleType.ORDINAL);
+      });
+    });
+
+    [TimeUnit.YEAR, TimeUnit.DATE, TimeUnit.MINUTES, TimeUnit.SECONDS, TimeUnit.MILLISECONDS, TimeUnit.YEARMONTH,
+     TimeUnit.YEARMONTHDATE, TimeUnit.YEARMONTHDATEHOURS, TimeUnit.YEARMONTHDATEHOURSMINUTES, TimeUnit.YEARMONTHDATEHOURSMINUTESSECONDS,
+     TimeUnit.HOURSMINUTES, TimeUnit.HOURSMINUTESSECONDS, TimeUnit.MINUTESSECONDS, TimeUnit.SECONDSMILLISECONDS,
+     TimeUnit.YEARQUARTER, TimeUnit.QUARTERMONTH, TimeUnit.YEARQUARTERMONTH].forEach((timeU) => {
```
Comment (kanitw) [inline test/query/encoding.test.ts]: timeU....

File: src/constraint/spec.ts [hunk 25]
```diff
@@ -325,8 +325,8 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
       for (let encQ of encodings) {
         if((encQ.channel === Channel.X || encQ.channel === Channel.Y) && encQ.scale) {
 
-          let sType = scaleType((encQ.scale as ScaleQuery).type as ScaleType,
-                                 encQ.timeUnit as TimeUnit, encQ.type as Type);
+          let sType = scaleType((encQ.scale as ScaleQuery).type,
```
Comment (kanitw) [inline src/constraint/spec.ts]: Why still leaving `as TimeUnit`?
